src/lib.rs

```diff
@@ -1363,43 +1419,12 @@ impl<T: Clone> Stream for Receiver<T> {
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         loop {
-            // If this stream is listening for events, first wait for a notification.
-            if let Some(listener) = self.listener.as_mut() {
-                ready!(Pin::new(listener).poll(cx));
-                self.listener = None;
-            }
-
-            loop {
-                // Attempt to receive a message.
-                match self.try_recv() {
-                    Ok(msg) => {
-                        // The stream is not blocked on an event - drop the listener.
-                        self.listener = None;
-                        return Poll::Ready(Some(msg));
-                    }
-                    Err(TryRecvError::Closed) => {
-                        // The stream is not blocked on an event - drop the listener.
-                        self.listener = None;
-                        return Poll::Ready(None);
-                    }
-                    Err(TryRecvError::Overflowed(_)) => continue,
-                    Err(TryRecvError::Empty) => {}
-                }
-
-                // Receiving failed - now start listening for notifications or wait for one.
-                match self.listener.as_mut() {
-                    None => {
-                        // Start listening and then try receiving again.
-                        self.listener = {
-                            let inner = self.inner.write().unwrap();
-                            Some(inner.recv_ops.listen())
-                        };
-                    }
-                    Some(_) => {
-                        // Go back to the outer loop to poll the listener.
-                        break;
-                    }
-                }
+            match ready!(self.as_mut().poll_recv(cx)) {
+                Some(Ok(val)) => return Poll::Ready(Some(val)),
+                // RecvError::Closed is handled upstream, so shouldn't be seen here.
+                None | Some(Err(RecvError::Closed)) => return Poll::Ready(None),
+                // If overflowed, we expect future operations to succeed so try again.
+                Some(Err(RecvError::Overflowed(_))) => continue,
```

Comment:
I'm not sure I follow how this changes anything. The `Stream` impl is still ignoring the overflow. :thinking:

Reply:
Yup; the idea in this PR was to not change the public Stream impl at all, but to add a way that we could poll the `Receiver<T>`, such that if I'm using `async-broadcast` internally in some struct `Foo`, I can write my own Stream impl for `Foo` which can internally poll the `async_broadcast::Receiver` and take advantage of this overflow information.

Reply:
that's is correct but `poll_recv` above is public and one could use that one to implement a custom stream impl with Overflow included...

Reply:
Ah ok, I get it now. Yeah, this is not disruptive at all and mostly just makes most of the inner code public so I don't have any objection to this approach.